src/coord/timestamp.rs

```diff
@@ -326,19 +367,73 @@ impl Timestamper {
     }
 
     /// Implements the byo timestamping logic
+    ///
+    /// If the partition count remains the same:
+    /// A new timestamp should be
+    /// 1) strictly greater than the last timestamp in this partition
+    /// 2) greater or equal to all the timestamps that have been assigned so far across all partitions
+    /// If the partition count increases:
+    /// A new timestamp should be:
+    /// 1) strictly greater than the last timestamp
+    /// This is necessary to guarantee that this timestamp *could not have been closed yet*
     fn update_byo_timestamp(&mut self) {
         for (id, byo_consumer) in &mut self.byo_sources {
             // Get the next set of messages from the Consistency topic
             let messages = byo_query_source(byo_consumer, self.max_increment_size);
             // Notify coordinator of updates
-            for (timestamp, offset) in byo_extract_ts_update(byo_consumer, messages) {
-                self.tx
-                    .unbounded_send(coord::Message::AdvanceSourceTimestamp {
-                        id: *id,
-                        timestamp,
-                        offset,
-                    })
-                    .expect("Failed to send update to coordinator");
+            for (partition_count, partition, timestamp, offset) in
+                byo_extract_ts_update(byo_consumer, messages)
+            {
+                let last_p_ts = match byo_consumer.last_partition_ts.get(&partition) {
+                    Some(ts) => *ts,
+                    None => 0,
+                };
+                if timestamp == 0
```

Comment:
consider separating this if statement into several if statements each with its own associated error message, and separately I don't think you need an else branch because taking the conditional takes you out of the function (I might be misreading here if so apologies)

Reply:
I like to avoid the else to avoid having another nested scope (this is partly scar tissue / ptsd from C where tabs are 8 spaces :) )